src/v2/Components/Sticky/Sticky.tsx

```diff
@@ -1,99 +1,252 @@
 import styled from "styled-components"
-import { Box } from "@artsy/palette"
-import React, { useEffect, useRef, useState } from "react"
+import { Box, useIsomorphicLayoutEffect } from "@artsy/palette"
+import React, {
+  cloneElement,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from "react"
 import { useSticky } from "./StickyProvider"
 import { useNavBarHeight } from "../NavBar/useNavBarHeight"
 
+interface StickyProps {
+  /**
+   * Maintains the original dimensions and horizontal position on page.
+   * Useful for sidebars.
+   **/
+  proportional?: boolean
+  /**
+   * Sticks only within the bounds of the containing element.
+   * Useful for split screen editorial.
+   **/
+  contained?: boolean
+}
+
+interface Rect {
+  right: number
+  left: number
+  width: number
+  height: number
+}
+
+const DEFAULT_RECT: Rect = {
+  right: 0,
+  left: 0,
+  width: 0,
+  height: 0,
+}
+
+enum State {
+  Pending,
+  AtTop,
+  AtBottom,
+}
+
 /**
  * Wrap a component to have it stick below the main nav.
  * Use render props `{({ stuck }) => {}` to swap styles.
- * See the stories for examples.
- *
- * FAQ:
- * - Q: Why can't I simply use `position: sticky`?
- * - A: We use `overflow-x: hidden` on the main layout container so that we can
- *      utilize the `FullBleed` component. When you set `overflow: hidden` on
- *      any ancestor of your sticky element, this element becomes the scrolling
- *      container for your sticky element.
- *
- *      See: https://github.com/w3c/csswg-drafts/issues/865 for more detail.
- *      TLDR: `overflow: clip` solves this and support is growing.
- *
- * - Q: How can I maintain the original horizontal position of the element I
- *      want to be sticky?
- * - A: Currently this implementation does not inherit any horizontal positioning.
- *      Everything that is stuck is the full width of the screen. See the
- *      `GridExample` story for an example that works around this by re-wrapping
- *      everything in an AppContainer + HorizontalPadding. (This is non-ideal)
- *
- * - Q: How do I unstick an element once it reaches the end of its parent container?
- * - A: This is also not currently supported.
  */
-export const Sticky: React.FC = ({ children }) => {
-  const sentinelRef = useRef<HTMLDivElement | null>(null)
+export const Sticky: React.FC<StickyProps> = ({
```

Comment:
One Q is: should this go in Palette? Seems appropriately lib-ish?

Reply:
Well right now it's got hardcoded assumptions about our nav in Force. Those [can/will go away](https://github.com/artsy/force/pull/7778) but for now we'll have to keep it here.